### PR TITLE
[WIP] refactor xgboost4j to create jvm-packages

### PR DESCRIPTION
**Pull request by CodingCat**
This is a WIP PR to create jvm-packages for xgboost

The major goal of the PR is to refactor the structure of the current xgboost4j and add scala APIs...

- [x] adjust the project structure

- [x] clean the minor problems in current Java code 

- [x] add Scala APIs

- [ ] add test cases for Scala APIs

- [ ] add Scala examples

- [x] fix and clean coding style (may add something like scalastyle check)

---

#### New XGBoostJVM directory structure ####

image

  The classes are put in the same package, so that we can hide the classes which shall be not exposed as PUBLIC, e.g. JNI-related classes...

  The users only see Booster/DMatrix/IEvaluation/IObjective/XGBoost (the original Trainer class, which provides API of cross-validation, train, init model, load model, etc)

 #### Scala API ####

Scala API is developed as the wrapper of Java Implementation, we have  `ScalaXGBoost` object which maps to Java-made XGBoost, 

image

the methods in ScalaXGBoost returns `ScalaBooster`, which is the wrapper of `Booster` in Java

image

---

Minor TODOs:

### CODE REVIEW

**Comment by tqchen** [comment]
please make a brief description of dir structure you are proposing, as the changelist is a bit hard to follow in github due to file movements

**Comment by CodingCat** [comment]
I will post it tomorrow together with the actual code change

**Comment by CodingCat** [comment]
@tqchen  @terrytangyuan @javelinjs @yanqingmen I posted the draft design and premilinary implementation of new JVM APIs, 

looking forward to your feedbacks

**Comment by javelinjs** [comment]
Maybe consider to be compatible with jdk1.6/1.7 ? e.g. avoid using java 8 features like ```HashMap<String, Object> params = new HashMap<>();```

**Comment by CodingCat** [comment]
good point

**Comment by CodingCat** [comment]
@javelinjs , good point, will fix it

**Comment by tqchen** [comment]
@CodingCat  My recommendation is use the same name. e.g. Booster instead of ScalaBooster, and put it under a different namespace(package) xgboost.scala.xx

This making things feels more "native", we are not aiming at building wrappers, but a full fledge JVM package that works naturally with scala and java,

**Comment by CodingCat** [comment]
Hmmm...just to be sure, implement the core functionality in Java and call it in scala is good, you just concern that the user shall not deal with different names when using different languages?

**Comment by tqchen** [comment]
Yes, it would be great when we create a booster, it is called Booster instead of ScalaBooster, since Scala is implicitly assumed when we use scala.

**Comment by CodingCat** [comment]
Sure, i will handle it

**Comment by tqchen** [comment]
just for reference, checkout https://github.com/dmlc/xgboost/pull/897 
I have made a python API working on YARN. The JVM implementation of distributed version could follow this in general

**Comment by CodingCat** [comment]
sure, thx

File: Makefile [hunk 1]
```diff
@@ -84,7 +84,7 @@ $(DMLC_CORE)/libdmlc.a:
 $(RABIT)/lib/$(LIB_RABIT):
 	+ cd $(RABIT); make lib/$(LIB_RABIT); cd $(ROOTDIR)
 
-java: java/libxgboost4j.so
+jvm-packages: jvm-packages/lib/libxgboost4j.so
```
Comment (tqchen) [inline Makefile]: simply use jvm

Comment (CodingCat) [inline Makefile]: you mean target or directory?

File: tests/travis/run_test.sh [hunk 2]
```diff
@@ -73,10 +73,9 @@ fi
 
 if [ ${TASK} == "java_test" ]; then
     set -e
-    make java
-    cd java
+    make jvm-packages
+    cd jvm-packages
     ./create_wrap.sh
```
Comment (tqchen) [inline tests/travis/run_test.sh]: is create_wrap still needed? reanem create_wrap to create_jni.sh

File: Makefile (hunk 1 above)
Comment (tqchen) [inline Makefile]: I mean target, this is a phony abbrv, so simple name should be fine
